sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -138,46 +138,80 @@ case class Not(child: Expression)
 case class In(value: Expression, list: Seq[Expression]) extends Predicate {
 
   require(list != null, "list should not be null")
+
+  lazy val valExprs = value match {
+    case cns: CreateNamedStruct => cns.valExprs
+    case expr => Seq(expr)
+  }
+
+  override lazy val resolved: Boolean = {
+    lazy val checkForInSubquery = list match {
+      case (l @ ListQuery(sub, children, _)) :: Nil =>
+        // SPARK-21759:
+        // It is possibly that the subquery plan has more output than value expressions, because
+        // the condition expressions in `ListQuery` might use part of subquery plan's output.
```

Comment:
This needs a better description. Instead of documenting the fact, this is caused by adding missing correlated predicate references when we pull up the correlated predicates. cc @dilipbiswal